### ISSUE TITLE
Enhance NTP configuration to support DNS and not just IP

### ISSUE THREAD

**Issue by ghormann** [FalconChristmas/fpp#1958]
**Is your feature request related to a problem? Please describe.**
Currently, NTP configuration only really supports IP address.  It seems to work if you provide a single server ip, but it doesn't work if you provide a pool URL.
![image](https://github.com/user-attachments/assets/88df6a44-1dd3-408f-9c86-e8b8dab9637e)

**Describe the solution you'd like**
- If a DNS lookup returns muitple IPs, configure as a "pool" record.
- If a single server DNS record is found or an IP address is provided, configure with "server"
- If left blank configure to use the falcon dns pool (no change vs today) 
- 
**Describe alternatives you've considered**
- Manually editing the file (discouraged)
- Only using Ip addresses, but then you can only provide one and not a different pool

**Additional context**
This was copied from #1956 and turned from a defect to an enhancement request.

**Comment by darylc** [comment]
A more foolproof method of detecting pools would be to do a dns lookup of the name and count the number of A records returned.

**Comment by ghormann** [comment]
> A more foolproof method of detecting pools would be to do a dns lookup of the name and count the number of A records returned.

Generally that probably is true onthe pulblic internet. (though I work for a company where that doesn't fully work).  Still, I updated the post.

**Comment by bloojoop** [comment]
for the record, the NTP fiels supports dns just fine and always has... I can validate and share that if you need me to... I do recoginize that is using a pool name, then it will find a single IP and not the full pool.

**Comment by bloojoop** [comment]
OK sooo. in 7.5 and earlier I was able to use a ntp server name... I have now discovered that any configured ntp server in v8.0 does not actually see to adjust the time.  I am using a direct IP address and in the troubleshooting commands. It will show the offset and delay and it is getting a query response from ntpq but it does not adjust the set the time initially.  

![image](https://github.com/user-attachments/assets/80447f9f-cf19-4086-b259-27158e9566b4)
![image](https://github.com/user-attachments/assets/731e720b-da6f-4c2c-ab54-3e2d77b17cd0)

I manually set the date and set the time to within 2 min and the offset shows better but still won't actually adjust. 
![image](https://github.com/user-attachments/assets/407e807a-c7ec-4959-97b6-494c5951d0fa)

**Comment by ghormann** [comment]
>... but it does not adjust the set the time initially.

@bloojoop  Am I correct you mean it doesn't adjust the time exactly **on boot**?  

I see your reach is 77 with a poll of 64, so this was taken after  6 ntp attempts. It is possible the system didn't have enough confidence to change the clock yet, and if you would have waited a bit longer it would have changed. It can take up to 10 minutes under good networking conditions. Reach is explained here: https://www.satsignal.eu/ntp/reach.html

When ntpdate was deprecated we probably  should have added a onetime call to "ntpd -g" on boot.  I did a quick scan (not complete) I'm not not 100% sure this is in the code.

**Comment by darylc** [comment]
Yeah this is a bit of a mess, talking to Ron yesterday I said I'd revert my previous commit which I've just done.

**Comment by bloojoop** [comment]
So I have 2 instances that have uptime of 23:55 (almost 24 hours) and have
never updated the time.  If I swap in a 7.5 card, get immediate time sync
and update.  If I remove the entry from the ntp server field, then it gets
time right away.. Of course this sets the ntp.conf back to pool
falconplayer.pool.ntp.org... with anything set in the gui, it sets to
server <whatever>... If I set the GUI to falconplayer.pool.ntp.org it never
gets time updated.   When directly editing the ntp.conf, if I cannot get
any 'server' entry to set the time but If I set the 'pool' entry to
debian.pool.ntp.org, or ubuntu.pool.ntp.org they work exactly as expected
but If I edit it to any other pool like time.nist.gov, time.windows.com,
time.google.com, or any other I tried. then it does not work.  I have only
gotten it to sync and set the time using pools in pool.ntp.org.. at these
that has been my testing for several hours.   I just ran sudo ntpd -g and
it ran but did not update the time.

**Comment by bloojoop** [comment]
Doing some more work on this.  When configured for the default poot, ntpq
-p shows several entries and one entry was time.cloudflare.com.  It showed
a reach of 1 and a delay of 4.5
***@***.***:~ $ ntpq -p
     remote                               refid      st t when poll reach
delay   offset   jitter
===================================================================================================
 falconplayer.pool.ntp.org           .POOL.          16 p    -  256    0
0.0000   0.0000   0.0010
+owners.kjsl.com                     75.114.196.163   2 u  153  256    1
 49.0677  -1.1153   1.1054
 arm1.maxhost.io                     128.138.141.172  2 u   90  256    1
 41.5969  -4.2214   4.1087
-198.71.50.75                        216.239.35.4     2 u  153  256    1
 26.3414  -2.4932   0.6824
-dutch.arpnet.net                    129.6.15.30      2 u  153  256    1
 41.2821   4.0173   0.3651
+s2-b.time.mci1.us.rozint.net        42.20.202.230    2 u  127  256    1
 27.6765  -5.7527   0.6189
+time.cloudflare.com                 10.125.8.50      3 u  127  256    1
4.5027  -1.3658   2.4321
-207.246.98.111.vultrusercontent.com 209.253.210.115  2 u  127  256    1
 42.6611  -4.7878   0.2101
*23.150.40.242                       204.9.54.119     2 u  127  256    1
 28.1127  -2.6521   0.6534

I then used time.cloudflare.com configured in the gui and got quite
different result with a reach of 377 and delay of 10.6
***@***.***:~ $ ntpq -p
     remote                               refid      st t when poll reach
delay   offset   jitter
===================================================================================================
+time.cloudflare.com                 10.125.8.50      3 u   51   64  377
 10.6625 6071402008   2.541

**Comment by ghormann** [comment]
So I did some testing by setting the date on my device to more than 1 month off.   I tried multiple configurations of ntp.conf.  What I found was that I **EITHER**  have to configure a pool, or I have to configure at least 3 NTP servers  before it would make a large time jump (e.g. 'panicgate'.

```txt
server 192.168.1.1 iburst minpoll 4 maxpoll 12 prefer
server time.windows.com iburst minpoll 4 maxpoll 12 prefer
server time.cloudflare.com iburst minpoll 4 maxpoll 12 prefer
```

After discovering this, I found  the quote 
```To work properly ntpd needs to talk to at least 3 servers ("A man with a watch knows what time it is. A man with two watches is never sure").``` -- https://www.ntppool.org/en/join/configuration.html#:~:text=To%20work%20properly%20ntpd%20needs,no%20more%20than%207%20servers.

The options I see are 
1) We go though a lot of tricks to use the `ntpd -q` option and hope this gets us close enough so that can survivle on a single ntp server
2) We check that the DNS is a pool, and if not prevent it from being used.
3) We change the UI allow multiple NTP servers and force at least 3.

**Comment by darylc** [comment]
What about option 4, remove the custom NTP setting.  It seems super unlikely that someone running an FPP that can't access the internet has access to 3 different NTP servers locally.  If they do then they are likely a super advanced user and can edit the config file.

**Comment by cathelest** [comment]
Is there a option to have a time server hosted on a fpp instance,

Eg. On a master pi that is connected via eth to the show router and wifi to another router for internet,

But other pi dont have internet access could set the time server to the master pi,

**Comment by cybercop23** [comment]
Why will the other Pis need access? Aren't they remotes being ran by the Player ( no more master)
Why not just set up masquerading/nat for the show network side so they all can get out and do what they want...

**Comment by ghormann** [comment]
> What about option 4, remove the custom NTP setting. It seems super unlikely that someone running an FPP that can't access the internet has access to 3 different NTP servers locally. If they do then they are likely a super advanced user and can edit the config file.

That is a very good point .

> The challenge with 1) ntpd -q, is that it does it once off and doesn't help with clock drift and it gives the user a false sense of security of NTP working when it's really just doing a once off sync.

From what I can tell (and I could be wrong), if the clock is within a second of the timesync server, 1 server will still work.     Thus, if ntpd -g is used to get it very close, then a single server can keep up with the drift.    

That said, personally I would either remove it or go with the "only pools allowed" option.

**Comment by darylc** [comment]
You are probably correct Greg, I did not test that scenario.  I propose we remove the custom setting in time for the 8.1 release (whenever that is), its a big improvement compared to the current situation in 8.x.  I'm not specifically against adding custom pools, but what does it actually gain end-users?  Use slightly closer NTP servers maybe? Unless I'm missing something, super marginal gain for the underlying FPP complexity of having the setting.

**Comment by bloojoop** [comment]
I will just add that, yes, multiple time servers is a best practice and for
time critical systems, it is required.  However, I have a LOT of other gear
that uses NTP and they all work just fine as expected with a single time
server configured.  This includes but is not limited to cisco networking
gear, netapp storage gear, vmware esx servers, ubuiquiti network gear
Synology NAS storage gear. This worked just fine with a single ntp server
with the  previous ntpd client in older OS.

If the best answer is to remove the setting, I can support that but
something is definitely different than before.  When doing my own digging
on ntp issues with bookworm, it appears that a LOT of the information I
find is around using systemd.timesyncd.  Is there a reason we are not using
systemd.timesyncd and using a different ntp client ?  just curious.

Ron.

**Comment by dkulp** [comment]
Just historical as we've always just used ntpd so all the scripts and stuff are setup for it.   If someone can figure out the systemd stuff and get it working, I'd definitely be OK with flipping over to it.

**Comment by dkulp** [comment]
Actually, systemd-timesync doesn't implement the time server part.  Thus, we could not use it in the case where one FPP instance acts as the time server for the others.   chrony might be more appropriate.

**Comment by bloojoop** [comment]
setting the minsane to 1 in my ntp.conf worked for me.

**Comment by ghormann** [comment]
Closed via fd44048
